synapse/handlers/stats.py

```diff
@@ -52,4 +59,324 @@ def __init__(self, hs):
+                # reduce the leave count when a new-to-the-room user joins.
+                prev_membership = None
+                if prev_event_id is not None:
+                    prev_event = yield self.store.get_event(
+                        prev_event_id, allow_none=True
+                    )
+                    if prev_event:
+                        prev_event_content = prev_event.content
+                        prev_membership = prev_event_content.get(
+                            "membership", Membership.LEAVE
+                        )
+
+                membership = event_content.get("membership", Membership.LEAVE)
+
+                if prev_membership is None:
+                    logger.debug("No previous membership for this user.")
+                elif prev_membership == Membership.JOIN:
+                    room_stats_delta["joined_members"] = (
+                        room_stats_delta.get("joined_members", 0) - 1
+                    )
+                elif prev_membership == Membership.INVITE:
+                    room_stats_delta["invited_members"] = (
+                        room_stats_delta.get("invited_members", 0) - 1
+                    )
+                elif prev_membership == Membership.LEAVE:
+                    room_stats_delta["left_members"] = (
+                        room_stats_delta.get("left_members", 0) - 1
+                    )
+                elif prev_membership == Membership.BAN:
+                    room_stats_delta["banned_members"] = (
+                        room_stats_delta.get("banned_members", 0) - 1
+                    )
+                else:
+                    err = "%s is not a valid prev_membership" % (repr(prev_membership),)
+                    logger.error(err)
+                    raise ValueError(err)
+
+                if membership == Membership.JOIN:
+                    room_stats_delta["joined_members"] = (
+                        room_stats_delta.get("joined_members", 0) + 1
+                    )
+                elif membership == Membership.INVITE:
+                    room_stats_delta["invited_members"] = (
+                        room_stats_delta.get("invited_members", 0) + 1
+                    )
+                elif membership == Membership.LEAVE:
+                    room_stats_delta["left_members"] = (
+                        room_stats_delta.get("left_members", 0) + 1
+                    )
+                elif membership == Membership.BAN:
+                    room_stats_delta["banned_members"] = (
+                        room_stats_delta.get("banned_members", 0) + 1
+                    )
+                else:
+                    err = "%s is not a valid membership" % (repr(membership),)
+                    logger.error(err)
+                    raise ValueError(err)
+
+                user_id = state_key
+                if self.is_mine_id(user_id) and membership in (
+                    Membership.JOIN,
+                    Membership.LEAVE,
```

Comment:
Why not the other memberships? You can go from `ban` to `leave` for example, or `join` to `ban` etc.